File: src/cpp/src/operations/tensor_invariants.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// von Mises / principal / hydrostatic / deviatoric invariants of a symmetric
// (6-component) or general (9-component row-major) tensor data array. See
// operations/tensor_invariants.hpp for the contract; the eigensolver lives in
// detail/sym3_eigen.hpp, shared with the CalculiX .frd reader's derived= option.

// System includes
#include <cmath>
#include <cstddef>
#include <limits>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

// Project includes
#include "meshioplusplus/operations/tensor_invariants.hpp"
#include "meshioplusplus/detail/data_ops.hpp"
#include "meshioplusplus/detail/sym3_eigen.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/ndarray.hpp"
#include "meshioplusplus/operations/data_common.hpp"
#include "meshioplusplus/parallel.hpp"

namespace meshioplusplus {

namespace {

/// Reads row @p r (0-based) of a 6- or 9-component array into the
/// `xx yy zz xy yz zx` order `detail::sym3_*` expects, symmetrizing the
/// off-diagonal terms for a 9-component input. Returns false (sym left
/// untouched) when any source component is non-finite.
bool tinv_read_symmetric(const NDArray& rArray, std::size_t Ncomp, std::size_t r, double* pSym) {
    double t[9];
    for (std::size_t k = 0; k < Ncomp; ++k)
        t[k] = detail::read_double(rArray, r * Ncomp + k);
    for (std::size_t k = 0; k < Ncomp; ++k)
        if (!std::isfinite(t[k]))
            return false;
    if (Ncomp == 6) {
        for (int k = 0; k < 6; ++k)
            pSym[k] = t[k];
        return true;
    }
    // 9-component row-major: xx xy xz yx yy yz zx zy zz.
    pSym[0] = t[0];
    pSym[1] = t[4];
    pSym[2] = t[8];
    pSym[3] = 0.5 * (t[1] + t[3]);
    pSym[4] = 0.5 * (t[5] + t[7]);
    pSym[5] = 0.5 * (t[2] + t[6]);
    return true;
}

/// The (unsymmetrized) hydrostatic mean of the tensor's own diagonal.
double tinv_hydrostatic(const NDArray& rArray, std::size_t Ncomp, std::size_t r) {
    const double xx = detail::read_double(rArray, r * Ncomp + 0);
    const double yy = detail::read_double(rArray, r * Ncomp + (Ncomp == 6 ? 1 : 4));
    const double zz = detail::read_double(rArray, r * Ncomp + (Ncomp == 6 ? 2 : 8));
    if (!std::isfinite(xx) || !std::isfinite(yy) || !std::isfinite(zz))
        return std::numeric_limits<double>::quiet_NaN();
    return (xx + yy + zz) / 3.0;
}

/// Number of rows of a Ncomp-wide array.
std::size_t tinv_nrows(const NDArray& rArray, std::size_t Ncomp) {
    return Ncomp > 0 ? rArray.Size() / Ncomp : 0;
}

NDArray tinv_mises(const NDArray& rSrc, std::size_t Ncomp) {
    const std::size_t nrows = tinv_nrows(rSrc, Ncomp);
    NDArray dst(DType::Float64, {nrows});
    double* out = dst.As<double>();
    parallel_for(nrows, [&](std::size_t r) {
        double sym[6];
        out[r] = tinv_read_symmetric(rSrc, Ncomp, r, sym)
                     ? detail::sym3_mises(sym)
                     : std::numeric_limits<double>::quiet_NaN();
    });
    return dst;
}

NDArray tinv_principal(const NDArray& rSrc, std::size_t Ncomp) {
    const std::size_t nrows = tinv_nrows(rSrc, Ncomp);
    NDArray dst(DType::Float64, {nrows, std::size_t{3}});
    double* out = dst.As<double>();
    parallel_for(nrows, [&](std::size_t r) {
        double sym[6];
        if (tinv_read_symmetric(rSrc, Ncomp, r, sym))
            detail::sym3_principal(sym, out + r * 3);
        else
            for (int k = 0; k < 3; ++k)
                out[r * 3 + k] = std::numeric_limits<double>::quiet_NaN();
    });
    return dst;
}

NDArray tinv_hydrostatic_array(const NDArray& rSrc, std::size_t Ncomp) {
    const std::size_t nrows = tinv_nrows(rSrc, Ncomp);
    NDArray dst(DType::Float64, {nrows});
    double* out = dst.As<double>();
    parallel_for(nrows, [&](std::size_t r) { out[r] = tinv_hydrostatic(rSrc, Ncomp, r); });
    return dst;
}

NDArray tinv_deviatoric(const NDArray& rSrc, std::size_t Ncomp) {
    const std::size_t nrows = tinv_nrows(rSrc, Ncomp);
    NDArray dst = NDArray::Uninit(DType::Float64, rSrc.Shape());
    double* out = dst.As<double>();
    const std::size_t yy_idx = Ncomp == 6 ? 1 : 4;
    const std::size_t zz_idx = Ncomp == 6 ? 2 : 8;
    parallel_for(nrows, [&](std::size_t r) {
        const double h = tinv_hydrostatic(rSrc, Ncomp, r);
        for (std::size_t k = 0; k < Ncomp; ++k) {
            const double v = detail::read_double(rSrc, r * Ncomp + k);
            const bool diag = (k == 0 || k == yy_idx || k == zz_idx);
            out[r * Ncomp + k] = (diag && std::isfinite(v) && std::isfinite(h)) ? v - h : v;
        }
    });
    return dst;
}

/// Whether an array's trailing component count is one `tensor_invariants` can
/// process.
bool tinv_is_tensor_shaped(const NDArray& rArray) {
    const std::size_t nc = data_num_components(rArray);
    return nc == 6 || nc == 9;
}

}  // namespace

TensorInvariant tensor_invariant_from_name(const std::string& rName) {
    unsigned mask = 0;
    std::size_t start = 0;
    bool any = false;
    while (start <= rName.size()) {
        const std::size_t comma = rName.find(',', start);
        const std::string tok = rName.substr(start, comma == std::string::npos ? std::string::npos
                                                                                : comma - start);
        if (tok == "mises")
            mask |= static_cast<unsigned>(TensorInvariant::Mises);
        else if (tok == "principal")
            mask |= static_cast<unsigned>(TensorInvariant::Principal);
        else if (tok == "hydrostatic")
            mask |= static_cast<unsigned>(TensorInvariant::Hydrostatic);
        else if (tok == "deviatoric")
            mask |= static_cast<unsigned>(TensorInvariant::Deviatoric);
        else if (tok == "all")
            mask |= static_cast<unsigned>(TensorInvariant::All);
        else
            throw std::invalid_argument(
                "meshio++: unknown tensor invariant '" + tok +
                "' (expected 'mises', 'principal', 'hydrostatic', 'deviatoric' or 'all')");
        any = true;
        if (comma == std::string::npos)
            break;
        start = comma + 1;
    }
    if (!any)
        throw std::invalid_argument("meshio++: tensor_invariant_from_name: empty list");
    return static_cast<TensorInvariant>(mask);
}

Mesh tensor_invariants(const Mesh& rMesh, const TensorInvariantsOptions& rOpts) {
    if (rOpts.location == DataLocation::Field)
        throw std::invalid_argument(
            "meshio++: tensor_invariants: field_data has no per-row tensor to reduce");

    std::vector<std::string> names = rOpts.names;
    if (names.empty()) {
        for (const std::string& n : data_names(rMesh, rOpts.location)) {
            const NDArray& src = rOpts.location == DataLocation::Point ? rMesh.PointData(n)
                                                                        : rMesh.CellData(n, 0);
            if (rOpts.location == DataLocation::Cell && rMesh.CellDataNumBlocks(n) == 0)
                continue;
            if (tinv_is_tensor_shaped(src))
                names.push_back(n);
        }
    } else {
        for (const std::string& n : names) {
            if (!data_has(rMesh, rOpts.location, n))
                throw std::invalid_argument(data_unknown_key_message(rMesh, rOpts.location, n));
            const NDArray& src = rOpts.location == DataLocation::Point ? rMesh.PointData(n)
                                                                        : rMesh.CellData(n, 0);
            if (!tinv_is_tensor_shaped(src))
                throw std::invalid_argument(
                    "meshio++: tensor_invariants: '" + n + "' has " +
                    std::to_string(data_num_components(src)) +
                    " component(s); expected 6 (symmetric) or 9 (general 3x3)");
        }
    }

    // Only ever called for the Point branch below (Field is rejected up front,
    // Cell has its own per-block helper), so it only needs AddPointData.
    auto add_point = [&](Mesh& rOut, const std::string& rTarget, NDArray Value) {
        if (!rOpts.overwrite && data_has(rOut, DataLocation::Point, rTarget))
            throw std::invalid_argument("meshio++: tensor_invariants: '" + rTarget +
                                        "' already exists (overwrite=false)");
        rOut.AddPointData(rTarget, std::move(Value));
    };

    Mesh out = detail::clone_mesh(rMesh);

    for (const std::string& name : names) {
        const std::string base = rOpts.prefix + name;

        if (rOpts.location == DataLocation::Cell) {
            const std::size_t nblocks = rMesh.NumCellBlocks();
            if (rMesh.CellDataNumBlocks(name) != nblocks)
                throw std::invalid_argument(
                    "meshio++: tensor_invariants: cell_data '" + name + "' has " +
                    std::to_string(rMesh.CellDataNumBlocks(name)) + " block(s) but the mesh has " +
                    std::to_string(nblocks) + " cell block(s)");
            if (nblocks == 0)
                continue;
            const std::size_t ncomp = data_num_components(rMesh.CellData(name, 0));

            auto add_cell = [&](const std::string& rSuffix,
                                NDArray (*pFn)(const NDArray&, std::size_t)) {
                const std::string target = base + rSuffix + rOpts.suffix;
                if (!rOpts.overwrite && data_has(out, DataLocation::Cell, target))
                    throw std::invalid_argument("meshio++: tensor_invariants: '" + target +
                                                "' already exists (overwrite=false)");
                std::vector<NDArray> blocks;
                blocks.reserve(nblocks);
                for (std::size_t b = 0; b < nblocks; ++b)
                    blocks.push_back(pFn(rMesh.CellData(name, b), ncomp));
                out.AddCellData(target, std::move(blocks));
            };
            if ((rOpts.outputs & TensorInvariant::Mises) != 0)
                add_cell("_mises", tinv_mises);
            if ((rOpts.outputs & TensorInvariant::Principal) != 0)
                add_cell("_principal", tinv_principal);
            if ((rOpts.outputs & TensorInvariant::Hydrostatic) != 0)
                add_cell("_hydrostatic", tinv_hydrostatic_array);
            if ((rOpts.outputs & TensorInvariant::Deviatoric) != 0)
                add_cell("_deviatoric", tinv_deviatoric);
            continue;
        }

        const NDArray& src = rMesh.PointData(name);
        const std::size_t ncomp = data_num_components(src);
        if ((rOpts.outputs & TensorInvariant::Mises) != 0)
            add_point(out, base + "_mises" + rOpts.suffix, tinv_mises(src, ncomp));
        if ((rOpts.outputs & TensorInvariant::Principal) != 0)
            add_point(out, base + "_principal" + rOpts.suffix, tinv_principal(src, ncomp));
        if ((rOpts.outputs & TensorInvariant::Hydrostatic) != 0)
            add_point(out, base + "_hydrostatic" + rOpts.suffix, tinv_hydrostatic_array(src, ncomp));
        if ((rOpts.outputs & TensorInvariant::Deviatoric) != 0)
            add_point(out, base + "_deviatoric" + rOpts.suffix, tinv_deviatoric(src, ncomp));
    }
    return out;
}

}  // namespace meshioplusplus

File: src/python/meshioplusplus/_tensor_invariants.py
"""``tensor_invariants``: von Mises, principal, hydrostatic and deviatoric
fields of a symmetric (6-component) or general 3x3 (9-component) tensor array.

A dependency-free mesh *operation* (not a file format). Before this operation
the only route to these was the CalculiX ``.frd`` reader's format-specific
``derived=`` option (:mod:`meshioplusplus.frd`); it now calls
:func:`tensor_invariants` internally and keeps its existing output names.

Input layout, per row:

- 6 components: a symmetric tensor ``xx yy zz xy yz zx`` (the order used
  throughout this codebase; see ``doc/mesh_data_model.md``).
- 9 components: a general 3x3 tensor, row-major (``xx xy xz yx yy yz zx zy
  zz``), the layout :func:`meshioplusplus.gradient`/:func:`meshioplusplus.hessian`
  produce. ``mises`` and ``principal`` are only defined for a symmetric
  tensor, so both use the symmetric part ``0.5*(T + T^T)``.

A row with any non-finite input component produces NaN in every requested
output for that row.

Public API:
    tensor_invariants
"""

from __future__ import annotations

import copy

import numpy as np

from ._data_common import location_map, normalize_location, num_components, require_key
from ._fallback import core_op_declined

_ALL_OUTPUTS = ("mises", "principal", "hydrostatic", "deviatoric")


def _symmetric6(flat, ncomp):
    """Rows reduced to ``xx yy zz xy yz zx``, symmetrizing a 9-comp input."""
    if ncomp == 6:
        return flat
    # Row-major 3x3: xx xy xz yx yy yz zx zy zz.
    xx, xy, xz, yx, yy, yz, zx, zy, zz = (flat[:, k] for k in range(9))
    return np.stack(
        [xx, yy, zz, 0.5 * (xy + yx), 0.5 * (yz + zy), 0.5 * (xz + zx)], axis=1
    )


def _mises(sym6):
    xx, yy, zz, xy, yz, xz = (sym6[:, k] for k in range(6))
    return np.sqrt(
        0.5
        * (
            (xx - yy) ** 2
            + (yy - zz) ** 2
            + (zz - xx) ** 2
            + 6.0 * (xy**2 + yz**2 + xz**2)
        )
    )


def _principal(sym6):
    """Eigenvalues of the symmetric tensors, ascending (min, mid, max)."""
    out = np.full((len(sym6), 3), np.nan)
    ok = np.isfinite(sym6).all(axis=1)
    if ok.any():
        s = sym6[ok]
        m = np.empty((len(s), 3, 3))
        m[:, 0, 0], m[:, 1, 1], m[:, 2, 2] = s[:, 0], s[:, 1], s[:, 2]
        m[:, 0, 1] = m[:, 1, 0] = s[:, 3]
        m[:, 1, 2] = m[:, 2, 1] = s[:, 4]
        m[:, 0, 2] = m[:, 2, 0] = s[:, 5]
        out[ok] = np.linalg.eigvalsh(m)
    return out


def _hydrostatic(flat, ncomp):
    """Mean of the tensor's own (unsymmetrized) diagonal."""
    yy_idx = 1 if ncomp == 6 else 4
    zz_idx = 2 if ncomp == 6 else 8
    xx, yy, zz = flat[:, 0], flat[:, yy_idx], flat[:, zz_idx]
    with np.errstate(invalid="ignore"):
        return (xx + yy + zz) / 3.0


def _deviatoric(flat, ncomp, hydro):
    yy_idx = 1 if ncomp == 6 else 4
    zz_idx = 2 if ncomp == 6 else 8
    out = flat.copy()
    ok = np.isfinite(hydro)
    for idx in (0, yy_idx, zz_idx):
        out[:, idx] = np.where(
            ok, flat[:, idx] - np.where(ok, hydro, 0.0), flat[:, idx]
        )
    return out


def _tensor_invariants_py(mesh, location, names, outputs, prefix, suffix, overwrite):
    """Pure-Python reference for :func:`tensor_invariants`."""
    loc = normalize_location(location)
    if loc == "field_data":
        raise ValueError(
            "meshio++: tensor_invariants: field_data has no per-row tensor to reduce"
        )
    source = location_map(mesh, loc)
    outs = list(outputs) if outputs else list(_ALL_OUTPUTS)
    for o in outs:
        if o not in _ALL_OUTPUTS:
            raise ValueError(
                f"meshio++: unknown tensor invariant '{o}' (expected one of "
                f"{', '.join(_ALL_OUTPUTS)})"
            )

    if names:
        selected = list(names)
        for name in selected:
            require_key(mesh, loc, name)
            ncomp = num_components(
                source[name][0] if loc == "cell_data" else source[name]
            )
            if ncomp not in (6, 9):
                raise ValueError(
                    f"meshio++: tensor_invariants: '{name}' has {ncomp} "
                    "component(s); expected 6 (symmetric) or 9 (general 3x3)"
                )
    else:
        selected = [
            name
            for name in sorted(source.keys())
            if num_components(source[name][0] if loc == "cell_data" else source[name])
            in (6, 9)
        ]

    out = copy.deepcopy(mesh)

    def _store(target, loc_, blocks_or_array):
        if not overwrite and target in location_map(out, loc_):
            raise ValueError(
                f"meshio++: tensor_invariants: '{target}' already exists (overwrite=false)"
            )
        location_map(out, loc_)[target] = blocks_or_array

    for name in selected:
        base = prefix + name
        if loc == "cell_data":
            blocks = source[name]
            if len(blocks) != len(mesh.cells):
                raise ValueError(
                    f"meshio++: tensor_invariants: cell_data '{name}' has "
                    f"{len(blocks)} block(s) but the mesh has "
                    f"{len(mesh.cells)} cell block(s)"
                )
            ncomp = num_components(blocks[0])
            per_output = {o: [] for o in outs}
            for b in blocks:
                flat = np.asarray(b, dtype=float).reshape(-1, ncomp)
                sym6 = _symmetric6(flat, ncomp)
                hydro = (
                    _hydrostatic(flat, ncomp)
                    if "hydrostatic" in outs or "deviatoric" in outs
                    else None
                )
                if "mises" in outs:
                    per_output["mises"].append(_mises(sym6))
                if "principal" in outs:
                    per_output["principal"].append(_principal(sym6))
                if "hydrostatic" in outs:
                    per_output["hydrostatic"].append(hydro)
                if "deviatoric" in outs:
                    per_output["deviatoric"].append(
                        _deviatoric(flat, ncomp, hydro).reshape(np.asarray(b).shape)
                    )
            for o in outs:
                _store(base + f"_{o}" + suffix, "cell_data", per_output[o])
            continue

        arr = np.asarray(source[name], dtype=float)
        ncomp = num_components(arr)
        flat = arr.reshape(-1, ncomp)
        sym6 = _symmetric6(flat, ncomp)
        if "mises" in outs:
            _store(base + "_mises" + suffix, loc, _mises(sym6))
        if "principal" in outs:
            _store(base + "_principal" + suffix, loc, _principal(sym6))
        hydro = None
        if "hydrostatic" in outs or "deviatoric" in outs:
            hydro = _hydrostatic(flat, ncomp)
        if "hydrostatic" in outs:
            _store(base + "_hydrostatic" + suffix, loc, hydro)
        if "deviatoric" in outs:
            _store(
                base + "_deviatoric" + suffix,
                loc,
                _deviatoric(flat, ncomp, hydro).reshape(arr.shape),
            )
    return out


def tensor_invariants(
    mesh,
    location: str = "point",
    keys=None,
    outputs=None,
    prefix: str = "",
    suffix: str = "",
    overwrite: bool = True,
):
    """Compute von Mises / principal / hydrostatic / deviatoric fields.

    Args:
        mesh: the source mesh (unmodified).
        location: ``"point"`` or ``"cell"`` (``"field"`` is rejected).
        keys: names to process; ``None`` means every 6- or 9-component array
            at ``location``.
        outputs: any of ``"mises"``, ``"principal"``, ``"hydrostatic"``,
            ``"deviatoric"``; ``None`` means all four.
        prefix: prepended to every output array's name.
        suffix: appended after the invariant's own name segment.
        overwrite: throw instead of silently overwriting an existing array of
            the target name.

    Returns:
        A new mesh with the source data plus the requested invariant arrays;
        geometry is unchanged.
    """
    loc = normalize_location(location)
    names = list(keys) if keys else []
    outs = list(outputs) if outputs else []
    args = (loc, names, outs, prefix, suffix, overwrite)
    try:
        from . import _core
    except ImportError:
        return _tensor_invariants_py(mesh, *args)
    try:
        out = _core.tensor_invariants(
            mesh, loc, names, ",".join(outs), prefix, suffix, overwrite
        )
    except Exception as exc:
        if not core_op_declined(exc, "tensor_invariants"):
            raise
        return _tensor_invariants_py(mesh, *args)
    for attr in ("point_sets", "cell_sets"):
        value = getattr(mesh, attr, None)
        if value:
            setattr(out, attr, copy.deepcopy(value))
    return out
